From 9c3926752b941c22361685b5482b35693c573813 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Fri, 7 Mar 2014 14:34:55 +0000
Subject: [PATCH 060/102] java: Use correct Set*Field JNI accessors to set
 fields of the appropriate type.

Using the wrong accessors (somehow - I have no idea how) caused other
fields in the struct to contain incorrect values.

(cherry picked from commit dff35285e491d1905d38c90d83fed98932cbc3b3)
---
 generator/java.ml | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/generator/java.ml b/generator/java.ml
index b07173b..d86304a 100644
--- a/generator/java.ml
+++ b/generator/java.ml
@@ -1137,13 +1137,13 @@ and generate_java_struct_return typ jtyp cols =
         pr "  (*env)->SetLongField (env, jr, fl, r->%s);\n" name;
     | name, (FUInt32|FInt32) ->
         pr "  fl = (*env)->GetFieldID (env, cl, \"%s\", \"I\");\n" name;
-        pr "  (*env)->SetLongField (env, jr, fl, r->%s);\n" name;
+        pr "  (*env)->SetIntField (env, jr, fl, r->%s);\n" name;
     | name, FOptPercent ->
         pr "  fl = (*env)->GetFieldID (env, cl, \"%s\", \"F\");\n" name;
         pr "  (*env)->SetFloatField (env, jr, fl, r->%s);\n" name;
     | name, FChar ->
         pr "  fl = (*env)->GetFieldID (env, cl, \"%s\", \"C\");\n" name;
-        pr "  (*env)->SetLongField (env, jr, fl, r->%s);\n" name;
+        pr "  (*env)->SetCharField (env, jr, fl, r->%s);\n" name;
   ) cols;
   pr "  free (r);\n";
   pr "  return jr;\n"
@@ -1192,11 +1192,11 @@ and generate_java_struct_list_return typ jtyp cols =
       | FBytes|FUInt64|FInt64 ->
         pr "    (*env)->SetLongField (env, jfl, fl, r->val[i].%s);\n" name;
       | FUInt32|FInt32 ->
-        pr "    (*env)->SetLongField (env, jfl, fl, r->val[i].%s);\n" name;
+        pr "    (*env)->SetIntField (env, jfl, fl, r->val[i].%s);\n" name;
       | FOptPercent ->
         pr "    (*env)->SetFloatField (env, jfl, fl, r->val[i].%s);\n" name;
       | FChar ->
-        pr "    (*env)->SetLongField (env, jfl, fl, r->val[i].%s);\n" name;
+        pr "    (*env)->SetCharField (env, jfl, fl, r->val[i].%s);\n" name;
   ) cols;
   pr "\n";
   pr "    (*env)->SetObjectArrayElement (env, jr, i, jfl);\n";
-- 
1.9.3

